### PR TITLE
Added support and tests for lowLatency, endLowLatency, and threaded port generation.

### LINKED ISSUE #70: Support .isolate(), .lowLatency(), and .endLowLatency() methods.
Currently, if a streams application is submitted with some form of a DISTRIBUTED context, the default is to run each operator in a separate PE. This is also the default for SPL programs, and has caused issues where customers create an excessive number of processes on their systems.

With the Java Application API, we are going to change the default behavior -- by default, every operator will run in the same PE, separated by threaded ports. 

If a Java Application API developers wants to separate subsequent operations into a separate PE they can call .isolate(), which returns a reference to a TStream whose child streams will run in a separate PE. To put this in SPL terms, everything runs in the same partition colocation by default; calling .isolate() creates another partition colocation in which the TStream's children run.

Lastly, introducing threaded ports between each operator incurs higher latency, so there will be a second mode of processing called "low latency". Low latency processing indicates that all operators will be created on the same PE <b>without</b> threaded ports between each other. Low latency processing can be started by calling a "lowLatency()" method on a TStream. This returns a reference to a TStream whose child operations will not be separated by threaded ports. Default processing may be resumed by calling an "endLowLatency()" method on a child TStream.

To reiterate: there are going to be three methods added to TStream:
.lowLatency()
.endLowLatency()
.isolate()

There will be two types of processing:
default - in which every operation (transformation, filtering, ...) is on the same PE, separated by threaded ports.
low latency - in which every operation is on the same PE, but are not separated by threaded ports.

### LINKED ISSUE #96: Address TStream.isolate feedback
Address comments in push request #85

### PR DESCRIPTION
**Pull request by wmarshall484**
Fixes #96 
Fixes #70

### CODE REVIEW

**Comment by wmarshall484** [comment]
jenkins test this please

**Comment by wmarshall484** [comment]
jenkins test this please

File: java/src/com/ibm/streamsx/topology/generator/spl/GraphUtilities.java [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+                    starts.add((JSONObject) ops.get(k));
+                }
+            }
+        }
+        return starts;
+    }
+
+    public static ArrayList<JSONObject> findOperatorByKind(String opKind,
+            JSONObject graph) {
+        ArrayList<JSONObject> lowLatencyStarts = new ArrayList<JSONObject>();
```
Comment (ddebrunner) [inline java/src/com/ibm/streamsx/topology/generator/spl/GraphUtilities.java]: Seems the method is generic so the collection should be called something like kindOperators, not lowLatencyStarts

File: java/src/com/ibm/streamsx/topology/generator/spl/GraphUtilities.java [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+            JSONObject op = (JSONObject) (ops.get(k));
+            String kind = (String) op.get("kind");
+            if (kind != null && kind.equals(opKind)) {
+                lowLatencyStarts.add(op);
+            }
+        }
+        return lowLatencyStarts;
+    }
+
+    // Get the children of the operator. Should probably be re-worked.
```
Comment (ddebrunner) [inline java/src/com/ibm/streamsx/topology/generator/spl/GraphUtilities.java]: Need some more detailed comments on what is returned, all downstream operators or only immediate? Any special handling for marker operators? Same for getParent.

Maybe consider chnaging the names to up/downstream.

File: test/java/src/com/ibm/streamsx/topology/test/api/IsolateTest.java [hunk 3]
```diff
@@ -21,54 +23,87 @@
...
-        
-	// Jenkins seems to fail when running this with a distributed tester.
-	// Not sure why, but the tests pass on my local build. 
-        /*try{
-        StreamsContextFactory
-                .getStreamsContext(StreamsContext.Type.DISTRIBUTED_TESTER)
-                .submit(topology).get(90, TimeUnit.SECONDS);
-        } catch(Exception e){
+
+        // Jenkins seems to fail when running this with a distributed tester.
```
Comment (ddebrunner) [inline test/java/src/com/ibm/streamsx/topology/test/api/IsolateTest.java]: The issue is that the test needs to follow the standard pattern of extending TestTopology and then have 'assume' checks that make it only run in distributed. currently Jenkins is not running distributed as it doesn't yet have an instance setup.

File: test/java/src/com/ibm/streamsx/topology/test/api/LowLatencyTest.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+    
+    @SuppressWarnings("serial")
+    private static Function<String, String> getPEId(){
+        return new Function<String, String>(){
+            int counter = 0;
+            @Override
+                public String apply(String v) {
+                // TODO Auto-generated method stub
+                return ((BigInteger) PERuntime.getCurrentContext().getPE().getPEId()).toString();
```
Comment (ddebrunner) [inline test/java/src/com/ibm/streamsx/topology/test/api/LowLatencyTest.java]: Curious why a cast to BigInteger is needed. Also there's an unused count and a TODO comment.

File: test/java/src/com/ibm/streamsx/topology/test/api/IsolateTest.java [hunk 5]
```diff
@@ -21,54 +23,87 @@
...
+        TStream<String> ss5 = ss4.transform(getPEId(), String.class).isolate();
+        ss5.transform(getPEId(), String.class).print();
+
+        TStream<String> ss7 = ss3.transform(getPEId(), String.class);
+
+        StreamsContextFactory.getStreamsContext(StreamsContext.Type.TOOLKIT)
+                .submit(topology).get();
     }
 
     @Test(expected = IllegalStateException.class)
```
Comment (ddebrunner) [inline test/java/src/com/ibm/streamsx/topology/test/api/IsolateTest.java]: Some comments as to why this should throw an exception, it looks valid to me, or is it a case that is just not yet covered?

**Pull request merged**